shell-completion/zsh/_systemctl.in

```diff
@@ -451,13 +451,9 @@ done
 
 (( $+functions[_systemctl_unit_properties] )) ||
     _systemctl_unit_properties() {
-        if ( [[ ${+_sys_all_properties} -eq 0 ]] || _cache_invalid SYS_ALL_PROPERTIES$_sys_service_mgr ) ||
-               ! _retrieve_cache SYS_ALL_PROPERTIES$_sys_service_mgr;
-        then
-            _sys_all_properties=( ${${(M)${(f)"$({{ROOTLIBEXECDIR}}/systemd --dump-bus-properties)"}}} )
-            _store_cache SYS_ALL_PROPERTIES$_sys_service_mgr _sys_all_properties
-        fi
-        _values -s , "${_sys_all_properties[@]}"
+		local -a _sys_all_properties=( ${(f)"$({{ROOTLIBEXECDIR}}/systemd --no-pager --dump-bus-properties 2>/dev/null)"} )
+        _wanted systemd-unit-properties expl 'unit property' \
+			_values -s , "${_sys_all_properties[@]}"
```

Comment:
Hmm, `systemctl show -P <TAB>` doesn't complete properties, it completes unit names for me ;(
But that's probably a bug elsewhere. I agree that caching this doesn't seem useful.

Reply:
Yeah, I also find this annoying. Unfortunately, that one is a longstanding bug and not a trivial fix with the current design. Since the main _arguments call dispatches to a simple handler for each systemctl verb, those handlers would need to understand the option arguments as well.

Idiomatic usage of _arguments for commands of this type often involves a tree of calls, and it seems systemd tries to take this approach but falls short. Maybe see the pacman zsh completions [1] for reference implementation of this pattern. There are other potential designs but they would likely mean a bigger rework of the completions.

[1] https://gitlab.archlinux.org/pacman/pacman/-/blob/master/scripts/completion/zsh_completion.in

Reply:
I've added a commit which should (partially) address this. You can let me know if it works well for you, but I might prefer to put that one in a separate PR.